modules/imgproc/test/test_distancetransform.cpp

```diff
@@ -283,4 +283,28 @@ void CV_DisTransTest::prepare_to_validation( int /*test_case_idx*/ )
 
 TEST(Imgproc_DistanceTransform, accuracy) { CV_DisTransTest test; test.safe_run(); }
 
+BIGDATA_TEST(Imgproc_DistanceTransform, large_image_12218)
+{
+    int i,j;
+    Size size(20000, 8000);
+    Mat src(size, CV_8UC1, Scalar::all(128));
+    Mat ans(size, CV_32FC1, Scalar::all(8192));
+    Mat dst;
+    Mat labels;
+
+    distanceTransform(src, dst, labels, cv::DIST_L2, cv::DIST_MASK_3, DIST_LABEL_PIXEL);
+
+    int* _ans = ans.ptr<int>();
+    int* _dst = dst.ptr<int>();
+    int ansstep = (int)(ans.step/sizeof(_ans[0]));
+    int dststep = (int)(dst.step/sizeof(_dst[0]));
+    for( i = 0; i < size.height; i++ )
+    {
+        float* d1 = (float*)(_dst + i*dststep);
+        float* d2 = (float*)(_ans + i*ansstep);
```

Comment:
Don't use manual pointer arithmetic.

At least there is mess of int / float types:
> _dst = dst.ptr<int>
> (float*)(_dst + i*dststep)

These "row" pointer should be received like this:
```
const float* d1 = dst.ptr<const float>(i);
const float* d2 = ans.ptr<const float>(i);
```

Anyway, this heavy loop is just a duplicate of existed norm function. So while check can be replaced to this:
```
EXPECT_LE(cvtest::norm(dst, ans, NORM_INF), 1e-3);
```